File: scripts/building_pipeline/parse_annotations.py
# DEPENDENCIES =================================================================

import os
from pathlib import Path
import pandas as pd


# ENVIROMENT ===================================================================
SHARED_DATA_PATH = \
    Path(os.environ["USERPROFILE"]) \
        / "Merck Sharp & Dohme, Corp" \
        / "IT Global Data Science - 2020-DigitalPathologyDS-MRL" \
        / "data"

ANNOTAION_PATH = SHARED_DATA_PATH / "annotations_to_parse"

BRAINS_CUT_PATH = SHARED_DATA_PATH / "brains_cut"

IMAGE_SIZE = {"X" : 1400,
              "Y" : 900}

# PROCESS ALL CSV FILES ========================================================
files_to_parse = os.listdir(ANNOTAION_PATH)

for file in files_to_parse:

    pairing_csv_df = []

    # pairing_csv_df = {idx = numeric(),
    #                   img = character(),
    #                   anno = character()}

    pairing_index = 1

    loaded_csv = pd.read_csv(
        ANNOTAION_PATH / file,
        header = None,
        names = ["X", "Y", "X3", "X4", "X5", "X6", "X7"],
        index_col = None,
    )

    study_folder_name = file.split(sep = ".")[0].replace("_", "")

    # create directory for given study
    annotations_folder = Path(BRAINS_CUT_PATH) / study_folder_name / "annotations"
    annotations_folder.mkdir(parents=True, exist_ok=True)

    loaded_csv["layer_folder_name"] = loaded_csv["X5"].apply(lambda g: "_".join(g.split("_")[2:6]))
    loaded_csv.sort_values(by="layer_folder_name")

    images_list = os.listdir(BRAINS_CUT_PATH / study_folder_name / "images")

    images_coords = pd.DataFrame(
        {
            "image_file_name": images_list,
            "coords": [(x, y) for (y, x) in
                     [image_name.split(".")[0].split("xx")[1].split("_")
                      for image_name in
                      images_list]]
         }
    )

    annot_max_min_df = loaded_csv.groupby("layer_folder_name",
                                          group_keys=False).agg(
        x_min = ("X", "min"),
        x_max = ("X", "max"),
        y_min=("Y", "min"),
        y_max=("Y", "max"),
    )

    annot_folders = sorted(loaded_csv["layer_folder_name"].unique())

    unique_x3 = loaded_csv.\
        groupby("layer_folder_name").\
        apply(lambda df: df["X3"].unique())

    for annot in annot_folders:
        layer_folder = annotations_folder / annot
        layer_folder.mkdir(parents = True, exist_ok = True)

        # Check if there are at least and at max 1 brain for annotation
        images_coords["true_flag"] =  images_coords["coords"].apply(
                lambda df: int(df[0]) <= annot_max_min_df.loc[annot]["x_min"] and
                           int(df[0]) + IMAGE_SIZE["X"] >=  annot_max_min_df.loc[annot]["x_max"] and
                           int(df[1]) <= annot_max_min_df.loc[annot]["y_min"] and
                           int(df[1]) + IMAGE_SIZE["Y"] >= annot_max_min_df.loc[annot]["y_max"]
            )

        # Check if there are at least and at max 1 brain for annotation
        if images_coords["true_flag"].sum() > 1:
            raise Exception("More then one brain for this annotation")
        elif images_coords["true_flag"].sum() == 0:
            raise Exception("No brain for this annotation")

        # creating pairing.csv dataframe - from anotations
        pairing_csv_df.append({
            "idx" : pairing_index,
            "img" : images_coords.loc[images_coords["true_flag"] == True].iloc[0]["image_file_name"],
            "anno" : annot
        })

        pairing_index += 1

        # For each unique X3 annotaion index, write separé file
        for x3 in list(unique_x3.loc[unique_x3.index == annot].iloc[0]):
            # create directory for given layer
            layer_folder = Path(BRAINS_CUT_PATH) \
                           / study_folder_name \
                           / "annotations" \
                           / annot


            layer_folder.mkdir(parents=True, exist_ok=True)

            #pos / neg annotation
            pos_affix = loaded_csv[(loaded_csv["layer_folder_name"] == annot) &
                                   (loaded_csv["X3"] == x3)]["X4"]

            if pos_affix.unique().__len__() > 1 :
                raise Exception("There are more then 1 unique positive/negative" \
                                + "indicator of annotation")
            else:
                if int(pos_affix.unique()) == 1:
                    pos_affix_str = "pos"
                else:
                    pos_affix_str = "neg"

            data_2_write = loaded_csv[(loaded_csv["layer_folder_name"] == annot) &
                                      (loaded_csv["X3"] == x3)][["X","Y"]]

           # write csv
            pd.DataFrame(data_2_write).to_csv(layer_folder \
                                                / ("file_" + str(x3) + "_" + pos_affix_str + ".csv"),
                                              index=False,
                                              header = False)

    # Add images to the pairing csv file which does not have annotation
    set_A = images_list
    set_B = list(pd.DataFrame(pairing_csv_df)["img"])
    missing_pairing_images = [i for i in set_A + set_B
                              if i not in set_A or i not in set_B]

    pairing_csv_pd_df = pd.DataFrame(pairing_csv_df)

    missing_pairing_images_df = \
        pd.DataFrame({"idx" : range(pairing_csv_pd_df["idx"].max() + 1,
                                    pairing_csv_pd_df["idx"].max() + 1 +
                                    len(missing_pairing_images)),
                      "img" : missing_pairing_images,
                      "anno" : [""] * len(missing_pairing_images)
     })

    pairing_csv_final_df = pd.concat([pairing_csv_pd_df,
                                      missing_pairing_images_df])

    pairing_csv_pd_df.to_csv(Path(BRAINS_CUT_PATH) \
                             / study_folder_name \
                             / "pairing.csv",
                             index=False)

# TAKE CARE OF FOLDERS WITHOUT ANNOTAION FILE ==================================

images_folder_list = os.listdir(BRAINS_CUT_PATH)

parsed_annotations = list(map(lambda g: g.split(".")[0].replace("_", ""),
                              files_to_parse))


missing_folders = list(set(images_folder_list) - set(parsed_annotations))

for missing_folder in missing_folders:
    # create directory for given study
    missing_annot_folder = Path(BRAINS_CUT_PATH) / missing_folder / "annotations"
    missing_annot_folder.mkdir(parents=True, exist_ok=True)

    miss_folder_images_list = os.listdir(BRAINS_CUT_PATH \
                                         / missing_folder \
                                         / "images")

    missing_folder_pairing_images_df = \
        pd.DataFrame({"idx" : range(1, len(miss_folder_images_list) + 1),
                      "img" : miss_folder_images_list,
                      "anno" : [""] * len(miss_folder_images_list)
     })

    missing_folder_pairing_images_df.to_csv(Path(BRAINS_CUT_PATH) \
                             / missing_folder \
                             / "pairing.csv",
                             index=False)

File: scripts/R_scripts/parse_annotations.R
# DEPENDENCIES =================================================================
packages = c("readr", "stringr", "dplyr", "purrr", "tidyr",
             "ggplot2")
## Now load or install&load all
package.check <- lapply(
  packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE)
      library(x, character.only = TRUE)
    }
  }
)


# ENVIROMENT ===================================================================
SHARED_DATA_PATH <- file.path(Sys.getenv("USERPROFILE"),
                              "Merck Sharp & Dohme, Corp",
                              "IT Global Data Science - 2020-DigitalPathologyDS-MRL",
                              "data")

ANNOTAION_PATH <- file.path(SHARED_DATA_PATH, 
                            "annotations_to_parse")

BRAINS_CUT_PATH <- file.path(SHARED_DATA_PATH,
                             "brains_cut")


image_size = data.frame(X = 1400, 
                        Y = 900)


# PROCESS ALL CSV FILES ========================================================
files_to_parse = list.files(ANNOTAION_PATH)


for (file in files_to_parse) {
  
  pairing_csv_df = data.frame(idx = numeric(),
                              img = character(),
                              anno = character())
  
  pairing_index = 1
  
  loaded_csv = read_csv(file.path(ANNOTAION_PATH, file), 
                        col_names = FALSE)
 
  study_folder_name = str_replace(str_split(file, "\\.")[[1]][1], 
                            pattern = "_", 
                            replacement = "")
  
  # create directory for given study
  dir.create(file.path(BRAINS_CUT_PATH, 
                       study_folder_name, 
                       "annotations"), recursive = T, showWarnings = F)


  # split by folders
  process_csv =
    loaded_csv %>%
    group_by(X5) %>%
    nest() %>%
    mutate(X5_parse = str_split(X5, "_")) %>%
    mutate(layer_folder_name = paste(X5_parse[[1]][3:5], collapse = "_")) %>%
    ungroup() %>%
    select(-X5,
           -X5_parse) %>%
    arrange(layer_folder_name)


  # PAIRING CSV CREATION ===========
  
  images_list = list.files(file.path(BRAINS_CUT_PATH,
                                     study_folder_name,
                                     "images"))
  
  
  images_coords = 
    map(images_list, 
        ~str_split(str_split(str_split(.x, 
                                       "\\.")[[1]][1], 
                             "xx")[[1]][2], 
                   "_")[[1]] %>% 
          t() %>% 
          as.data.frame()) %>% 
    reduce(bind_rows) %>% 
    mutate(Y = as.numeric(V1), 
           X = as.numeric(V2)) %>% 
    select(-V1, 
           -V2) %>% 
    bind_cols(data.frame(image_file_name = images_list)) %>% 
    select(X, Y, image_file_name)
  
  
  annot_folders = 
    process_csv %>% 
    distinct(layer_folder_name) %>% 
    arrange(layer_folder_name)

  
  for (folder_index in 1:nrow(annot_folders)) {

    annot_singl_folder = pull(annot_folders[folder_index, 1])


    dir.create(file.path(BRAINS_CUT_PATH,
                         study_folder_name,
                         "annotations",
                         annot_singl_folder),
               recursive = T, 
               showWarnings = F)

    # annotations' coordinates
    data_per_folder =
      process_csv %>%
      filter(layer_folder_name == annot_singl_folder) %>%
      select(data) %>%
      unnest()

    
    # min and max of annotation
    annot_max_min_df = 
      data.frame(x_min = data_per_folder %>% select(X1) %>% map( ~min(.x)) %>% unlist(),
                 x_max = data_per_folder %>% select(X1) %>% map( ~max(.x)) %>% unlist(),
                 y_min = data_per_folder %>% select(X2) %>% map( ~min(.x)) %>% unlist(),
                 y_max = data_per_folder %>% select(X2) %>% map( ~max(.x)) %>% unlist())
             
     
    # find brain corresponding to this folder's annotation   
    correct_img_coords = 
      images_coords %>% 
      mutate(x_min_flag = ifelse(X <= annot_max_min_df$x_min, 1, 0),
             x_max_flag = ifelse(X + image_size$X >= annot_max_min_df$x_max, 1, 0),
             y_min_flag = ifelse(Y <= annot_max_min_df$y_min, 1, 0),
             y_max_flag = ifelse(Y + image_size$Y >= annot_max_min_df$y_max, 1, 0)) %>% 
      mutate(final_flag = x_min_flag * x_max_flag * y_min_flag * y_max_flag) %>% 
      filter(final_flag == 1) %>% 
      select(X, Y, image_file_name) %>% 
      mutate(image_file_name = as.character(image_file_name))
          
    # Check if there are at least and at max 1 brain for annotation       
    if (nrow(correct_img_coords) > 1) {
      stop("There are more then 1 brain with the same annotation")
    } else if (nrow(correct_img_coords) == 0) {
      stop("No brain for annotation found")
    }        
    
    
    pairing_csv_df = 
      pairing_csv_df %>% 
      bind_rows(data.frame(idx = pairing_index,
                           img = as.character(correct_img_coords$image_file_name[1]),
                           anno = as.character(annot_singl_folder)))
    
    pairing_index = pairing_index + 1
         
    # Writting the file_X_[pos/neg].csv into correct folder
    X3_index =
      data_per_folder %>%
      distinct(X3)

    for (indx in 1:nrow(X3_index)) {
      file_anot_name = pull(X3_index[indx, 1])

      file_name = file.path(BRAINS_CUT_PATH,
                            study_folder_name,
                            "annotations",
                            annot_singl_folder,
                            paste0("file_",
                                   file_anot_name,
                                   "_",
                                   ifelse(data_per_folder %>% 
                                            filter(X3 == file_anot_name) %>% 
                                            head(1) %>% 
                                            pull(X4)  == 1,
                                          "pos",
                                          "neg"),
                                   ".csv"))

      data_per_folder %>%
        filter(X3 == file_anot_name) %>% 
        select(X1, X2) %>%
        write_csv(file_name,
                  col_names = F)

    }

  }
 
  pairing_csv_df = 
    pairing_csv_df %>% 
    bind_rows(
      data.frame(idx = seq(from = max(pairing_csv_df$idx) + 1, 
                           length.out = length(setdiff(images_list, pairing_csv_df$img))), 
                 img = setdiff(images_list, pairing_csv_df$img), 
                 anno = character(length(setdiff(images_list, pairing_csv_df$img)))))
  
  # Write down the pairing.csv file into study_folder
  write_csv(pairing_csv_df,
            file.path(BRAINS_CUT_PATH, 
                      study_folder_name, 
                      "pairing.csv"), 
            col_names = T)
}


# TAKE CARE OF FOLDERS WITHOUT ANNOTAION FILE ==================================


images_folder_list = list.files(file.path(BRAINS_CUT_PATH))


parsed_annotations = 
  files_to_parse %>% 
  map(~str_replace(str_split(.x, "\\.")[[1]][1], 
            pattern = "_", 
            replacement = "")) %>% 
  unlist()


missing_folders = setdiff(images_folder_list, parsed_annotations)

for (miss_folder_idx in 1:length(missing_folders)) {
  
  miss_folder = missing_folders[miss_folder_idx]
  
  # create directory for given study
  dir.create(file.path(BRAINS_CUT_PATH, 
                       miss_folder, 
                       "annotations"), recursive = T, showWarnings = F)
  
  images_list = list.files(file.path(BRAINS_CUT_PATH,
                                     miss_folder,
                                     "images"))
  
  pairing_csv_df = 
    data.frame(idx = seq(from = 1, 
                         length.out = length(images_list)), 
               img = images_list, 
               anno = character(length(images_list)))
  
  write_csv(pairing_csv_df,
            file.path(BRAINS_CUT_PATH, 
                      miss_folder, 
                      "pairing.csv"), 
            col_names = T)
  
}
